ractor_safe_table.c

```diff
@@ -0,0 +1,323 @@
+                RUBY_ASSERT(hash == table->funcs->hash(key));
+                inserting = true;
+            }
+
+            rb_atomic_t prev_size = RUBY_ATOMIC_FETCH_ADD(table->size, 1);
+
+            if (UNLIKELY(prev_size > table->capacity / 2)) {
+                ractor_safe_table_try_resize(table_obj, table_obj_ptr);
+
+                goto retry;
+            }
+
+            curr_key = RUBY_ATOMIC_VALUE_CAS(entry->key, RACTOR_SAFE_TABLE_EMPTY, key);
+            if (curr_key == RACTOR_SAFE_TABLE_EMPTY) {
+                RUBY_ATOMIC_VALUE_SET(entry->hash, hash);
+
+                RB_GC_GUARD(table_obj);
+                return key;
+            }
+            else {
+                // Entry was not inserted.
+                RUBY_ATOMIC_DEC(table->size);
+
+                // Another thread won the race, try again at the same location.
+                continue;
+            }
+          }
+          case RACTOR_SAFE_TABLE_DELETED:
+            break;
+          case RACTOR_SAFE_TABLE_MOVED:
+            // Wait
+            RB_VM_LOCKING();
+
+            goto retry;
+          default: {
+            VALUE curr_hash = RUBY_ATOMIC_VALUE_LOAD(entry->hash);
+            if ((curr_hash == hash || curr_hash == 0) && table->funcs->cmp(key, curr_key)) {
+                // We've found a match.
+                if (UNLIKELY(rb_objspace_garbage_object_p(curr_key))) {
+                    // This is a weakref table, so after marking but before sweeping is complete we may find a matching garbage object.
+                    // Skip it and mark it as deleted.
+                    RUBY_ATOMIC_VALUE_CAS(entry->key, curr_key, RACTOR_SAFE_TABLE_DELETED);
+
+                    // Fall through and continue our search.
+                }
+                else {
+                    RB_GC_GUARD(table_obj);
+                    return curr_key;
+                }
+            }
+
+            break;
+          }
+        }
+
+        idx = ractor_safe_table_probe_next(&probe);
+    }
+}
+
+VALUE
+rb_ractor_safe_table_delete(VALUE table_obj, VALUE key)
+{
+    // Assume locking and barrier (which there is no assert for).
+    ASSERT_vm_locking();
+
+    struct ractor_safe_table *table = RTYPEDDATA_GET_DATA(table_obj);
+
+    VALUE hash = table->funcs->hash(key);
+
+    struct ractor_safe_table_probe probe;
+    int idx = ractor_safe_table_probe_start(&probe, table, hash);
+
+    while (true) {
+        struct ractor_safe_table_entry *entry = &table->entries[idx];
+        VALUE curr_key = RUBY_ATOMIC_VALUE_LOAD(entry->key);
+
+        switch (curr_key) {
+          case RACTOR_SAFE_TABLE_EMPTY:
+            // We didn't find our entry to delete.
+            return 0;
+          case RACTOR_SAFE_TABLE_DELETED:
+            break;
+          case RACTOR_SAFE_TABLE_MOVED:
+            rb_bug("rb_ractor_safe_table_delete: moved entry");
+            break;
+          default:
+            if (table->funcs->cmp(key, curr_key)) {
```

Comment:
In the existing code we compared by identity here, right? I'd really like to keep that as otherwise it's easy to end up with [Bug #21172](https://bugs.ruby-lang.org/issues/21172)

Reply:
Maybe we could call this `rb_ractor_safe_table_delete_by_identity`? I feel like we're going to want to do this for the symbol table as well (since we're likely to build duplicate dsymbols when concurrently inserting).